File: frontend/src/lib/triage.ts
import type { IconName } from "../components/brand";
import type { ServiceType, SymptomType } from "./api";

/**
 * Mirror of the server's triage table (backend/app/services/triage.py).
 * The client keeps it so the follow-up question can be shown without a round
 * trip; the server is the source of truth for the mapping at request time.
 */

export const SYMPTOMS: {
  id: SymptomType;
  label: string;
  blurb: string;
  icon: IconName;
  tint?: "fuel" | "electrical" | "tyre" | "recovery";
  hero?: boolean;
  wide?: boolean;
}[] = [
  {
    id: "out_of_fuel",
    label: "Out of fuel",
    blurb: "Empty tank, gauge on E",
    icon: "nozzle",
    tint: "fuel",
    hero: true,
  },
  { id: "wont_start", label: "Won't start", blurb: "No response, no crank", icon: "battery", tint: "electrical" },
  { id: "flat_tyre", label: "Flat tyre", blurb: "Tyre deflated, can't drive", icon: "tyre", tint: "tyre" },
  { id: "cant_move", label: "Car won't move", blurb: "Stuck in place, won't drive", icon: "tow", tint: "recovery" },
  { id: "locked_out", label: "Locked out", blurb: "Keys stuck inside", icon: "key" },
  { id: "something_else", label: "Something else", blurb: "Tell us what's wrong", icon: "dots", wide: true },
];

const BRANCHES: Partial<Record<SymptomType, string>> = {
  wont_start: "Do the lights come on?",
  flat_tyre: "Do you have a spare wheel?",
};

const ANSWER_TO_SERVICE: Record<string, ServiceType> = {
  lights_yes: "mechanic",
  lights_no: "jump_start",
  spare_yes: "tyre_change",
  spare_no: "towing",
};

const DIRECT: Partial<Record<SymptomType, ServiceType>> = {
  out_of_fuel: "fuel",
  cant_move: "towing",
  locked_out: "lockout",
  something_else: "mechanic",
};

export function followUpQuestion(symptom: SymptomType | null): string | null {
  return symptom ? (BRANCHES[symptom] ?? null) : null;
}

export function resolveService(
  symptom: SymptomType | null,
  answer: string | null,
): ServiceType {
  if (symptom && DIRECT[symptom]) return DIRECT[symptom];
  if (symptom && BRANCHES[symptom]) return ANSWER_TO_SERVICE[answer ?? ""] ?? "mechanic";
  return "mechanic";
}

export type TriageStep = {
  symptom: SymptomType | null;
  answer: string | null;
  service: ServiceType;
};

File: backend/app/services/triage.py
"""Symptom triage: resolve the motorist's description to exactly one service.

The client keeps this table mirrored so the follow-up question can be shown
without a round trip; the server is the source of truth for the mapping. Both
the reported symptom AND the resolved service are persisted on the order so
dispatch accuracy can be measured later.
"""

from ..models import ServiceType, SymptomType

# Which follow-up question to ask for a symptom (a "branch"), keyed by answer.
# Answers are small machine tokens ("lights_yes", "spare_no", ...) so matching
# stays stable across the client/server boundary.
BRANCHES: dict[str, str] = {
    SymptomType.WONT_START: "Do the lights come on?",
    SymptomType.FLAT_TYRE: "Do you have a spare wheel?",
}

ANSWER_TO_SERVICE: dict[str, ServiceType] = {
    "lights_yes": ServiceType.MECHANIC,
    "lights_no": ServiceType.JUMP_START,
    "spare_yes": ServiceType.TYRE_CHANGE,
    "spare_no": ServiceType.TOWING,
}

# Symptoms with no follow-up resolve straight to a service.
DIRECT: dict[str, ServiceType] = {
    SymptomType.OUT_OF_FUEL: ServiceType.FUEL,
    SymptomType.CANT_MOVE: ServiceType.TOWING,
    SymptomType.LOCKED_OUT: ServiceType.LOCKOUT,
    SymptomType.SOMETHING_ELSE: ServiceType.MECHANIC,
}


def follow_up_question(symptom: str | None) -> str | None:
    if symptom is None:
        return None
    return BRANCHES.get(symptom)


def resolve_service(symptom: str | None, answer: str | None) -> ServiceType:
    """Symptom + answer → exactly one service type.

    Unknown or missing inputs fall back to generic assistance rather than
    erroring out mid-flow; the direct service grid sends no symptom at all.
    """
    if symptom in DIRECT:
        return DIRECT[symptom]
    if symptom in BRANCHES:
        return ANSWER_TO_SERVICE.get(answer or "", ServiceType.MECHANIC)
    return ServiceType.MECHANIC
